File: templates/project/nodejs/src/lakebase-credentials.js
// Runtime Lakebase credential minting for the app + knex migrations.
//
// No DB token is stored in .env. This module mints a short-lived Postgres
// credential ON DEMAND from the connection METADATA in the environment
// (LAKEBASE_PROJECT_ID, LAKEBASE_BRANCH_ID, LAKEBASE_ENDPOINT), caches it
// in-process, and re-mints before it expires. Mirrors the kit's
// get-connection.ts / mintCredential seam (the databricks CLI is the single
// credential source), so the pattern is identical across languages.

const { execFileSync } = require('node:child_process');

// Re-mint a token this many ms after it was minted. Lakebase database
// credentials live ~1h; 40 min keeps a comfortable margin so a pooled
// connection never presents an expired token.
const MINT_TTL_MS = 40 * 60 * 1000;
const DEFAULT_ENDPOINT = 'primary';

let cachedToken = null;
let mintedAt = 0;

// Build the Lakebase endpoint resource path from env METADATA, or null when the
// metadata needed to mint is absent (so a caller can fall back).
function endpointPathFromEnv() {
  const instance = process.env.LAKEBASE_PROJECT_ID;
  const branch = process.env.LAKEBASE_BRANCH_ID;
  if (!instance || !branch) return null;
  const endpoint = process.env.LAKEBASE_ENDPOINT || DEFAULT_ENDPOINT;
  return `projects/${instance}/branches/${branch}/endpoints/${endpoint}`;
}

function profileArgs() {
  const p = process.env.DATABRICKS_CONFIG_PROFILE;
  return p ? ['--profile', p] : [];
}

function runDatabricks(args) {
  return execFileSync('databricks', [...args, ...profileArgs()], { encoding: 'utf8' });
}

// Return a valid Lakebase DB token, minting a fresh one via the databricks CLI
// when the cache is empty or near expiry.
function mintToken() {
  const endpoint = endpointPathFromEnv();
  if (!endpoint) {
    throw new Error(
      'Cannot mint a Lakebase credential: LAKEBASE_PROJECT_ID / LAKEBASE_BRANCH_ID ' +
        'are not set. The post-checkout hook sets them; otherwise provide DATABASE_URL explicitly.'
    );
  }
  if (cachedToken && Date.now() - mintedAt < MINT_TTL_MS) return cachedToken;
  const raw = runDatabricks(['postgres', 'generate-database-credential', endpoint, '-o', 'json']);
  const token = (JSON.parse(raw) || {}).token;
  if (!token) throw new Error(`generate-database-credential returned no token for ${endpoint}`);
  cachedToken = token;
  mintedAt = Date.now();
  return token;
}

// The Lakebase user (email): prefer the DB_USERNAME metadata, else the CLI.
function currentUser() {
  const u = process.env.DB_USERNAME;
  if (u) return u;
  const raw = runDatabricks(['current-user', 'me', '-o', 'json']);
  const me = JSON.parse(raw) || {};
  return me.userName || (me.emails && me.emails[0] && me.emails[0].value) || '';
}

module.exports = { mintToken, endpointPathFromEnv, currentUser };

File: templates/project/python/app/lakebase_credentials.py
"""Runtime Lakebase credential minting for the app + migrations.

No DB token is stored in .env. Instead the app mints a short-lived Postgres
credential ON DEMAND from the connection METADATA in the environment
(LAKEBASE_PROJECT_ID, LAKEBASE_BRANCH_ID, LAKEBASE_ENDPOINT), caches it
in-process, and re-mints before it expires. This mirrors the kit's
get-connection.ts / mintCredential seam (the databricks CLI is the single
credential source), so the pattern is identical across languages.
"""

from __future__ import annotations

import json
import os
import subprocess
import threading
import time

# Re-mint a token this many seconds after it was minted. Lakebase database
# credentials live ~1h; 40 min keeps a comfortable margin so a pooled
# connection never presents an expired token (paired with pool_recycle in
# database.py).
_MINT_TTL_SECONDS = 40 * 60

_DEFAULT_ENDPOINT = "primary"

_lock = threading.Lock()
_cached_token: str | None = None
_minted_at: float = 0.0


def endpoint_path_from_env() -> str | None:
    """Build the Lakebase endpoint resource path from env METADATA, or None when
    the metadata needed to mint is absent (so a caller can fall back)."""
    instance = os.getenv("LAKEBASE_PROJECT_ID")
    branch = os.getenv("LAKEBASE_BRANCH_ID")
    if not instance or not branch:
        return None
    endpoint = os.getenv("LAKEBASE_ENDPOINT", _DEFAULT_ENDPOINT)
    return f"projects/{instance}/branches/{branch}/endpoints/{endpoint}"


def _profile_args() -> list[str]:
    profile = os.getenv("DATABRICKS_CONFIG_PROFILE")
    return ["--profile", profile] if profile else []


def _run_databricks(args: list[str]) -> str:
    proc = subprocess.run(
        ["databricks", *args, *_profile_args()],
        capture_output=True,
        text=True,
        check=True,
    )
    return proc.stdout


def mint_token(*, force: bool = False) -> str:
    """Return a valid Lakebase DB token, minting a fresh one via the databricks
    CLI when the cache is empty or near expiry. Thread-safe + cached."""
    global _cached_token, _minted_at
    endpoint = endpoint_path_from_env()
    if endpoint is None:
        raise RuntimeError(
            "Cannot mint a Lakebase credential: LAKEBASE_PROJECT_ID / "
            "LAKEBASE_BRANCH_ID are not set. The post-checkout hook sets them; "
            "otherwise provide DATABASE_URL explicitly."
        )
    with _lock:
        fresh_enough = (
            _cached_token is not None
            and (time.monotonic() - _minted_at) < _MINT_TTL_SECONDS
        )
        if fresh_enough and not force:
            return _cached_token
        raw = _run_databricks(
            ["postgres", "generate-database-credential", endpoint, "-o", "json"]
        )
        token = (json.loads(raw) or {}).get("token")
        if not token:
            raise RuntimeError(
                f"generate-database-credential returned no token for {endpoint}"
            )
        _cached_token = token
        _minted_at = time.monotonic()
        return token


def current_user() -> str:
    """The Lakebase user (email). Prefer the DB_USERNAME metadata; fall back to
    `databricks current-user me`."""
    user = os.getenv("DB_USERNAME")
    if user:
        return user
    raw = _run_databricks(["current-user", "me", "-o", "json"])
    me = json.loads(raw) or {}
    emails = me.get("emails") or []
    return me.get("userName") or (emails[0].get("value") if emails else "") or ""
